### PR TITLE
gh-93911: Specialize `LOAD_ATTR` for custom `__getattr__` and `__getattribute__`

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Linked to #93911.

For `__getattr__`, we specialize as per normal as long as the specialized bytecode can succeed without raising `AttributeError`.

For `__getattribute__`, a specialized instruction is created to inline the call to `__getattribute__`.

* Issue: gh-93911

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
A very significant (15%) increase in hit rates for `test_typing`. That module represents the best case for this specialisation because nearly all its classes use `__getattr__`. This is also the [highest specialization failure](https://github.com/faster-cpython/ideas/blob/main/stats.md#load_attr-1) apart from `property` which was fixed by the other PR.
```
.\PCbuild\amd64\python_d.exe -m test test_typing -R 3:3

On main:
opcode[106].specializable : 1
    opcode[106].specialization.success : 3833
    opcode[106].specialization.failure : 3648
    opcode[106].specialization.hit : 861058
    opcode[106].specialization.deferred : 520385
    opcode[106].specialization.miss : 91899
    opcode[106].specialization.deopt : 1598
    opcode[106].execution_count : 119454
    opcode[106].specialization.failure_kinds[2] : 1918
    opcode[106].specialization.failure_kinds[4] : 214
    opcode[106].specialization.failure_kinds[5] : 9
    opcode[106].specialization.failure_kinds[9] : 171
    opcode[106].specialization.failure_kinds[11] : 251
    opcode[106].specialization.failure_kinds[12] : 207
    opcode[106].specialization.failure_kinds[14] : 148
    opcode[106].specialization.failure_kinds[17] : 201
    opcode[106].specialization.failure_kinds[19] : 22
    opcode[106].specialization.failure_kinds[21] : 5
    opcode[106].specialization.failure_kinds[22] : 14
    opcode[106].specialization.failure_kinds[23] : 39
    opcode[106].specialization.failure_kinds[25] : 18
    opcode[106].specialization.failure_kinds[27] : 464


On this PR:
opcode[106].specializable : 1
    opcode[106].specialization.success : 4694
    opcode[106].specialization.failure : 2298
    opcode[106].specialization.hit : 991871
    opcode[106].specialization.deferred : 355737
    opcode[106].specialization.miss : 128358
    opcode[106].specialization.deopt : 2267
    opcode[106].execution_count : 156019
    opcode[106].specialization.failure_kinds[2] : 46
    opcode[106].specialization.failure_kinds[4] : 214
    opcode[106].specialization.failure_kinds[5] : 9
    opcode[106].specialization.failure_kinds[9] : 171
    opcode[106].specialization.failure_kinds[11] : 251
    opcode[106].specialization.failure_kinds[12] : 207
    opcode[106].specialization.failure_kinds[14] : 155
    opcode[106].specialization.failure_kinds[17] : 201
    opcode[106].specialization.failure_kinds[19] : 22
    opcode[106].specialization.failure_kinds[21] : 5
    opcode[106].specialization.failure_kinds[22] : 13
    opcode[106].specialization.failure_kinds[23] : 39
    opcode[106].specialization.failure_kinds[25] : 18
    opcode[106].specialization.failure_kinds[27] : 464
    opcode[106].specialization.failure_kinds[28] : 516
```

File: Python/ceval.c [hunk 1]
```diff
@@ -3688,6 +3689,99 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+                _Py_call_attribute(owner, getattribute, name)
+                : PyObject_GenericGetAttr(owner, name);
+            if (res) {
+                SET_TOP(NULL);
+                STACK_GROW((oparg & 1));
+                SET_TOP(res);
+                Py_DECREF(owner);
+                JUMPBY(INLINE_CACHE_ENTRIES_LOAD_ATTR);
+                DISPATCH();
+            }
```
Comment (Fidget-Spinner) [inline Python/ceval.c]: I'm aware that specialized opcodes should have as few branches as possible. But these two opcodes need to respect the `getattro` semantics.

I could move these out to a function, but the branch would still be there.

**Comment by markshannon** [comment]
This doesn't seem right to me.
`__getattribute__` allows attribute lookup to be overridden.
`__getattr__` is for looking attributes *after* normal lookup has failed.

The existence of `__getattr__` shouldn't prevent specialization of normal attribute lookup.
Consider this class
```Python
>>> class C:
...     def __init__(self):
...         self.a = 1
...     b = 2
...     def __getattr__(self, name):
...         if name == "c":
...             return 3
...         raise AttributeError(name)
... 
>>> c = C()
>>> c.a
1
>>> c.b
2
>>> c.c
3
>>> c.d
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 8, in __getattr__
AttributeError: d
```
We should specialize `c.a` and `c.b`; the presence of `__getattr__` shouldn't matter.

Interestingly, the current behavior of Python is to check for both `__getattribute__` and `__getattr__` before calling either of them.
```Python
def get_b(self, name):
    print("In get_b")
    if name == "b":
        return "b"
    else:
        raise AttributeError()

class A:
    def __getattribute__(self, name):
        print("In A.__getattribute__")
        if name == "a":
            return "a"
        else:
            type(self).__getattr__ = get_b
            raise AttributeError()

try:
    print(A().b)
except Exception as e:
    print("Raises", type(e), e)
print()
try:
    print(A().b)
except Exception as e:
    print("Raises", type(e), e)
```
cause the following to be printed:
```
In A.__getattribute__
Raises <class 'AttributeError'> 

In A.__getattribute__
In get_b
b
```
Which is odd, but ideal for specialization, as we don't need to check for `__getattr__` at runtime. We can do it at specialization time.

**Comment by Fidget-Spinner** [comment]
Alright. `analyze_descriptor` will need to be a lot more robust then. It will need to detect cases where it sees a `__getattr__` but the attribute is present.

**Comment by Fidget-Spinner** [comment]
It's alarming that specialisation misses increased more than hits (~75k vs 55k). Maybe it means that for most types with `__getattr__`, it's expected that there's nothing there? This needs more analysis.
```
python -m test test_typing -R 3:3

Main 71868a0066a90519ccc6aeb75673ae882aaa03f0:
opcode[106].specializable : 1
    opcode[106].specialization.success : 3839
    opcode[106].specialization.failure : 3608
    opcode[106].specialization.hit : 892801
    opcode[106].specialization.deferred : 517029
    opcode[106].specialization.miss : 91724
    opcode[106].specialization.deopt : 1596
    opcode[106].execution_count : 119325
    opcode[106].specialization.failure_kinds[2] : 1918
    opcode[106].specialization.failure_kinds[4] : 213
    opcode[106].specialization.failure_kinds[5] : 9
    opcode[106].specialization.failure_kinds[9] : 172
    opcode[106].specialization.failure_kinds[11] : 254
    opcode[106].specialization.failure_kinds[12] : 165
    opcode[106].specialization.failure_kinds[14] : 148
    opcode[106].specialization.failure_kinds[17] : 203
    opcode[106].specialization.failure_kinds[19] : 20
    opcode[106].specialization.failure_kinds[21] : 5
    opcode[106].specialization.failure_kinds[22] : 14
    opcode[106].specialization.failure_kinds[23] : 39
    opcode[106].specialization.failure_kinds[25] : 18
    opcode[106].specialization.failure_kinds[27] : 463

This PR 2a931a8faad07375af647f5ed0e8d3548bcb3d37:
opcode[106].specializable : 1
    opcode[106].specialization.success : 5321
    opcode[106].specialization.failure : 2387
    opcode[106].specialization.hit : 944885
    opcode[106].specialization.deferred : 391933
    opcode[106].specialization.miss : 164726
    opcode[106].specialization.deopt : 2937
    opcode[106].execution_count : 192307
    opcode[106].specialization.failure_kinds[2] : 568
    opcode[106].specialization.failure_kinds[4] : 273
    opcode[106].specialization.failure_kinds[5] : 9
    opcode[106].specialization.failure_kinds[9] : 172
    opcode[106].specialization.failure_kinds[11] : 287
    opcode[106].specialization.failure_kinds[12] : 165
    opcode[106].specialization.failure_kinds[14] : 148
    opcode[106].specialization.failure_kinds[17] : 234
    opcode[106].specialization.failure_kinds[19] : 20
    opcode[106].specialization.failure_kinds[21] : 5
    opcode[106].specialization.failure_kinds[22] : 14
    opcode[106].specialization.failure_kinds[23] : 39
    opcode[106].specialization.failure_kinds[25] : 51
    opcode[106].specialization.failure_kinds[27] : 468
```

File: Python/ceval.c [hunk 2]
```diff
@@ -3681,6 +3682,47 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+            PyCodeObject *code = (PyCodeObject *)f->func_code;
+            assert(code->co_argcount == 2);
+            DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
+            STAT_INC(LOAD_ATTR, hit);
+
+            PyObject *name = GETITEM(names, oparg >> 1);
+            Py_INCREF(f);
+            _PyInterpreterFrame *new_frame = _PyFrame_PushUnchecked(tstate, f);
+            SET_TOP(NULL);
+            int push_null = !(oparg & 1);
```
Comment (markshannon) [inline Python/ceval.c]: We push `NULL` if `oparg & 1` (and the line below should be `STACK_SHRINK(1-push_null)`)

File: Python/specialize.c [hunk 3]
```diff
@@ -1212,6 +1263,34 @@ function_kind(PyCodeObject *code) {
     return SIMPLE_FUNCTION;
 }
 
+static uint32_t
+maybe_function_get_version(PyObject *o, int expected_argcount, int opcode)
```
Comment (markshannon) [inline Python/specialize.c]: Why the `maybe_` prefix? I would find this clearer as `get_function_version`.
Either way, please add a comment stating that returning zero is a failure.

File: Python/specialize.c [hunk 4]
```diff
@@ -791,6 +823,24 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name)
         case GETSET_OVERRIDDEN:
             SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_OVERRIDDEN);
             goto fail;
+        case GET_CALL_GETATTRIBUTE:
+        {
+            getattrofunc getattro_slot = type->tp_getattro;
```
Comment (markshannon) [inline Python/specialize.c]: Unused variable.

File: Python/specialize.c [hunk 5]
```diff
@@ -791,6 +823,24 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name)
...
+            assert(getattro_slot == _Py_slot_tp_getattro);
```
Comment (markshannon) [inline Python/specialize.c]: Avoid the unused variable warning with `assert(type->tp_getattro == _Py_slot_tp_getattro);`

File: Python/specialize.c [hunk 6]
```diff
@@ -1212,6 +1263,34 @@ function_kind(PyCodeObject *code) {
     return SIMPLE_FUNCTION;
 }
 
+static uint32_t
+maybe_function_get_version(PyObject *o, int expected_argcount, int opcode)
+{
+    uint32_t version = 0;
+    if (!Py_IS_TYPE(o, &PyFunction_Type)) {
```
Comment (markshannon) [inline Python/specialize.c]: This needs a `SPECIALIZATION_FAIL(...`

File: Python/specialize.c [hunk 7]
```diff
@@ -568,8 +574,47 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
     }
     else {
         if (type->tp_getattro != PyObject_GenericGetAttr) {
-            *descr = NULL;
-            return GETSET_OVERRIDDEN;
+            getattrofunc getattro_slot = type->tp_getattro;
```
Comment (markshannon) [inline Python/specialize.c]: I wonder if our assumption that `getattro_slot == _Py_slot_tp_getattr_hook` implies that there is a `__getattr__` is true.
It might be that the slot hasn't been updated, even though that seems unlikely.

Could you add an assert that `_PyType_Lookup(type, &_Py_ID(__getattr__)) != NULL` if `has_getattr`?

File: Include/internal/pycore_typeobject.h [hunk 8]
```diff
@@ -44,6 +44,9 @@ extern PyStatus _PyTypes_InitSlotDefs(void);
 extern void _PyStaticType_Dealloc(PyTypeObject *type);
 
 
+extern PyObject *_Py_slot_tp_getattro(PyObject *self, PyObject *name);
+extern PyObject *_Py_slot_tp_getattr_hook(PyObject *self, PyObject *name);
+extern PyObject *_Py_call_attribute(PyObject *self, PyObject *attr, PyObject *name);
```
Comment (markshannon) [inline Include/internal/pycore_typeobject.h]: Does this need to be externally visible?

**Comment by markshannon** [comment]
The numbers look correct. The total, `hit + deferred + miss` is unchanged and the number of deferred drops from 34% to 26%, which is a decent improvement. The increase in misses from 6% to 10% may be genuine, due to type instability. You are running this on the tests after all.

Have you tried to gather stats for pyperformance?

**Review by markshannon — CHANGES_REQUESTED** [review]
Looks good, but I'd like to assert our assumptions about `tp_getattro`.

File: Python/ceval.c (hunk 2 above)
Comment (Fidget-Spinner) [reply Python/ceval.c]: What about:
```
int shrink_stack = !(oparg & 1);
STACK_SHRINKG(shrink_stack);
```
The behaviour is equivalent but it does save a subtract operation.

File: Python/specialize.c (hunk 3 above)
Comment (Fidget-Spinner) [reply Python/specialize.c]: `maybe_` indicates this can accept a non-`PyFunctionObject` (in which case it fails and returns 0. But I can change the name since it doesn't seem to be helping clarity.

File: Python/specialize.c (hunk 7 above)
Comment (Fidget-Spinner) [reply Python/specialize.c]: `_Py_slot_tp_getattr_hook` handles this case by overwriting itself back to the simpler `_Py_slot_tp_getattro`. So I think that assert would fail. But in this case would it matter? If there's no `__getattr__` isn't that a good thing? It doesn't change the fact that we can just do normal attribute lookup.

File: Python/specialize.c [hunk 9]
```diff
@@ -1212,6 +1262,36 @@ function_kind(PyCodeObject *code) {
     return SIMPLE_FUNCTION;
 }
 
+/* Returning 0 indicates a failure. */
+static uint32_t
+function_get_version(PyObject *o, int expected_argcount, int opcode)
+{
+    uint32_t version = 0;
+    if (!Py_IS_TYPE(o, &PyFunction_Type)) {
```
Comment (markshannon) [inline Python/specialize.c]: This should go back to `case PROPERTY:` the `case GET_CALL_GETATTRIBUTE:` already asserts that `o` is a function.
Also the name `function_get_version` implies that `o` must be a function.

File: Python/specialize.c [hunk 10]
```diff
@@ -1212,6 +1262,36 @@ function_kind(PyCodeObject *code) {
     return SIMPLE_FUNCTION;
 }
 
+/* Returning 0 indicates a failure. */
+static uint32_t
+function_get_version(PyObject *o, int expected_argcount, int opcode)
```
Comment (markshannon) [inline Python/specialize.c]: The name of this function is misleading.
Could you split it into two functions,  `function_checkargs` and `function_get_version`.

File: Python/specialize.c [hunk 11]
```diff
@@ -792,6 +824,23 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name)
         case GETSET_OVERRIDDEN:
             SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_OVERRIDDEN);
             goto fail;
+        case GET_CALL_GETATTRIBUTE:
+        {
+            assert(type->tp_getattro == _Py_slot_tp_getattro);
+            assert(Py_IS_TYPE(descr, &PyFunction_Type));
+            _PyLoadMethodCache *lm_cache = (_PyLoadMethodCache *)(instr + 1);
+            uint32_t func_version = function_get_version(descr, 2,
```
Comment (markshannon) [inline Python/specialize.c]: See comment in `function_get_version`, but this does two things.
`uint32_t func_version = function_check_args(descr, 2) && function_get_version(descr)`

File: Python/specialize.c (hunk 7 above)
Comment (markshannon) [reply Python/specialize.c]: What are we checking for here?

This is what I think we want to do w.r.t overriding these methods:
* Neither `__getattribute__` nor `__getattr__` are overridden, which we already handle.
* Both are overridden. Rare, don't specialize.
* `__getattribute__` is overridden, `__getattr__` is not, and `__getattribute__` is a Python function, which we add a new specialization for.
* `__getattr__` is overridden, two cases:
   *  We expect a value on the instance. We can treat this as a normal instance lookup, as we de-opt if the value is `NULL`, falling back to the full lookup.
  * No attribute on the instance. This is going to be rare, so don't specialize.

Is that correct?

If so, could you set two booleans `has_getattr` and `has_getattribute`first, then do the case analysis, it would be clearer (to me at least).

File: Python/ceval.c [hunk 12]
```diff
@@ -3722,6 +3723,47 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
+            _PyLoadMethodCache *cache = (_PyLoadMethodCache *)next_instr;
+            PyObject *owner = TOP();
+            PyTypeObject *cls = Py_TYPE(owner);
+            uint32_t type_version = read_u32(cache->type_version);
+            DEOPT_IF(cls->tp_version_tag != type_version, LOAD_ATTR);
+            assert(type_version != 0);
+            PyObject *getattribute = read_obj(cache->descr);
+            assert(Py_IS_TYPE(getattribute, &PyFunction_Type));
+            PyFunctionObject *f = (PyFunctionObject *)getattribute;
+            uint32_t func_version = read_u32(cache->keys_version);
```
Comment (markshannon) [inline Python/ceval.c]: We don't need to check the function version, just that it takes exactly two arguments.

File: Python/specialize.c [hunk 13]
```diff
@@ -569,8 +575,47 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
...
+                &_Py_ID(__getattribute__));
+            bool custom_getattribute = !(getattribute == NULL ||
+                (Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
+                    ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
+                    (void *)PyObject_GenericGetAttr));
+            if (custom_getattribute) {
+                if (getattro_slot == _Py_slot_tp_getattro &&
+                    Py_IS_TYPE(getattribute, &PyFunction_Type)) {
+                    *descr = getattribute;
+                    return GET_CALL_GETATTRIBUTE;
```
Comment (markshannon) [inline Python/specialize.c]: `GETATTRIBUTE_IS_PYTHON_FUNCTION` maybe?

**Review by markshannon — CHANGES_REQUESTED** [review]
Looks like a worthwhile improvement.

A few suggestions.

File: Python/specialize.c (hunk 11 above)
Comment (Fidget-Spinner) [reply Python/specialize.c]: This is really smart! Thanks.

**Comment by Fidget-Spinner** [comment]
> What are we checking for here?
> ,,,
> This is what I think we want to do w.r.t overriding these methods:

> Is that correct?

Yes.

> If so, could you set two booleans has_getattr and has_getattributefirst, then do the case analysis, it would be clearer (to me at least).

There's one more special case for custom `__getattribute__`: where it's set to the defaut lookup (`PyObject_GenericGetAttr`). In such cases if there's no `__getattr__`, we can just treat this as a normal attributel ookup.

**Comment by markshannon** [comment]
> There's one more special case for custom __getattribute__: where it's set to the defaut lookup (PyObject_GenericGetAttr). In such cases if there's no __getattr__, we can just treat this as a normal attribute lookup. This is why it matters whether it's a custom getattribute or not.

Do you mean 
```Python
class C:
    __getattribute__ = object.__getattribute__
```
?

Isn't that the same as not overriding at all?

**Comment by Fidget-Spinner** [comment]
> > There's one more special case for custom **getattribute**: where it's set to the defaut lookup (PyObject_GenericGetAttr). In such cases if there's no **getattr**, we can just treat this as a normal attribute lookup. This is why it matters whether it's a custom getattribute or not.
> 
> Do you mean
> 
> ```python
> class C:
>     __getattribute__ = object.__getattribute__
> ```
> 
> ?
> 
> Isn't that the same as not overriding at all?

No I was trying to say that it is the same in that case :).

File: Python/specialize.c [hunk 14]
```diff
@@ -569,8 +577,47 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
...
+               __getattr__ and then installs the simpler slot if necessary. */
+            /* Using some custom tp_getattro, like super() objects */
+            if (getattro_slot != _Py_slot_tp_getattr_hook &&
+                getattro_slot != _Py_slot_tp_getattro) {
+                *descr = NULL;
+                return GETSET_OVERRIDDEN;
+            }
+            PyObject *getattribute = _PyType_Lookup(type,
+                &_Py_ID(__getattribute__));
+            has_custom_getattribute = getattribute != NULL &&
```
Comment (markshannon) [inline Python/specialize.c]: This seems convoluted, and I'm not convinced it is correct.
Would this work?
```C
PyObject *object_getattribute = PyType_Lookup(&PyBaseObject_Type,&_Py_ID(__getattribute__));
has_custom_getattribute = getattribute != object_getattribute;
```
We should probably cache the value of `object_getattribute`

Comment (Fidget-Spinner) [reply Python/specialize.c]: I copied it from what `slot_tp_getattro` uses https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7836-L7840.

So at the very least, it's consistent with how `typeobject.c` does it.

Comment (Fidget-Spinner) [reply Python/specialize.c]: After some thought, I'd prefer we keep it as-is. I'd rather follow the `typeobject.c` implementation closely than implement a more correct check which may deviate from how `typeobject.c` behaves.

Comment (markshannon) [reply Python/specialize.c]: That check is in `slot_tp_getattr_hook`, which only applies if `__getattr__` is set. 
What if `__getattr__` is not set?

Comment (Fidget-Spinner) [reply Python/specialize.c]: Then it calls `__getattribute__` https://github.com/python/cpython/blob/main/Objects/typeobject.c#L7781-L7784. In this case since there's no `__getattr__` and no custom `__getattribute__`, it just fails through and specializes to normal attribute lookup. 
```pycon
>>> class X:
...  __getattribute__ = object.__getattribute__
...  def __init__(s):
...   s.x = 1
...
>>> x = X()
>>>
>>> def f():
...  for _ in range(10):
...   x.x
...
>>> f()
>>> import dis
>>> dis.dis(f, adaptive=True)
  1           0 RESUME_QUICK             0

  2           2 LOAD_GLOBAL_ADAPTIVE     1 (NULL + range)
             14 LOAD_CONST               1 (10)
             16 CALL_ADAPTIVE            1
             26 GET_ITER
        >>   28 FOR_ITER_RANGE          19 (to 70)
             32 STORE_FAST               0 (_)

  3          34 LOAD_GLOBAL_MODULE       2 (x)
             46 LOAD_ATTR_INSTANCE_VALUE     2 (x)
             66 POP_TOP
             68 JUMP_BACKWARD_QUICK     21 (to 28)

  2     >>   70 LOAD_CONST               0 (None)
             72 RETURN_VALUE
>>>
```

Comment (markshannon) [reply Python/specialize.c]: You are copying code from `slot_tp_getattr_hook`, but that is only correct if `getattro_slot == _Py_slot_tp_getattr_hook` is set. 

Not only that, but there is no guarantee that class's version number changes when `type->tp_getattro` changes, so we cannot rely on the invariants implied by `slot_tp_getattr_hook`.

The `type->tp_getattro` dance is an implementation of the true semantics.
It is is independent of specialization and it may change in a way that is correct, but breaks your assumptions here.
For example, we may remove `PyWrapperDescr_Type` and `PyWrapperDescrObject` and replace them with builtin functions in future.

Comment (Fidget-Spinner) [reply Python/specialize.c]: > Not only that, but there is no guarantee that class's version number changes when type->tp_getattro changes, so we cannot rely on the invariants implied by slot_tp_getattr_hook

That's tough. You mean it could cycle between `_Py_slot_tp_getattr_hook` or `_Py_slot_tp_getattro` right? If it's just that we can check at runtime that things stay as expected.

If you mean that any arbitrary type could just modify its `type->tp_getattro` at runtime. That would mean all of our current `LOAD_ATTR` specialized opcodes will need runtime checking that `tp_getattro` stays as `PyObject_GenericGetAttr`. How do I know a C type doesn't change to a custom lookup function halfway through? However, the perf impact for such an unlikely case isn't worth it. I'd be willing to break such obscure code if it means slowing down Python overall by 1% just to support these.

> You are copying code from slot_tp_getattr_hook, but that is only correct if getattro_slot == _Py_slot_tp_getattr_hook is set.

I can add another check at specialization time that ``getattribute == object_getattribute`` like you suggested. But I don't see how the current (more permissive) check could be wrong either. If a wrapper descriptor wraps a `PyObject_GenericGetAttr`, how can its lookup be anything but standard?

Are you alright with me checking both, that is :
```C
            has_custom_getattribute = getattribute != NULL &&
                !((Py_IS_TYPE(getattribute, &PyWrapperDescr_Type) &&
                    ((PyWrapperDescrObject *)getattribute)->d_wrapped ==
                    (void *)PyObject_GenericGetAttr) ||
                    getattribute != object_getattribute);
```
?
When the time comes in the future when we change ``object.__getattribute__``. We can just change the specializing code.

Comment (markshannon) [reply Python/specialize.c]: What I mean is that we shouldn't rely on how `_Py_slot_tp_getattr_hook` or `_Py_slot_tp_getattro` behave.
If either are set, then we should look up `__getattribute__` and `__getattr__` directly and use those.
All we should rely on is:
* `type->tp_getattro == PyObject_GenericGetAttr`: Normal behaviour.
* `type->tp_getattro == _Py_slot_tp_getattr_hook` **or** `type->tp_getattro == _Py_slot_tp_getattro`: The `tp_getattro` hasn't been specified in the type struct, but either `__getattribute__` or `__getattr__` may have been overridden.

The first case is fine. We already handle it.
The second case is a bit more complex, but well defined [here](https://docs.python.org/3/reference/datamodel.html#customizing-attribute-access). Rather than trying to decipher the code in typeobject.c, we should follow the docs.

Comment (markshannon) [reply Python/specialize.c]: Would this work?
```C
bool has_getattr;
getattrofunc getattro_slot = type->tp_getattro;
if (getattro_slot == PyObject_GenericGetAttr) {
    /* Normal attribute lookup; */
   has_getattr = false;
}
else if (getattro_slot == _Py_slot_tp_getattro || getattro_slot == _Py_slot_tp_getattr_hook) {
    /* One or both of __getattribute__ or __getattr__ may have been overridden */ 
    PyObject *getattribute = _PyType_Lookup(type, &_Py_ID(__getattribute__));
    PyObject *getattr = _PyType_Lookup(type, &_Py_ID(__getattr__));
    bool has_custom_getattribute = getattribute != object_getattribute;
    has_getattr = getattr != NULL;
    if (has_custom_getattribute) {
        if (!has_getattr && Py_IS_TYPE(getattribute, &PyFunction_Type)) {
            *descr = getattribute;
            return GETATTRIBUTE_IS_PYTHON_FUNCTION;
        }
        *descr = NULL;
        return GETSET_OVERRIDDEN;
    }
}
else {
    /* Something else */
   *descr = NULL;
   return GETSET_OVERRIDDEN;
}
```

We could replace the logic for setting `has_getattr` with `has_getattr = getattro_slot == _Py_slot_tp_getattr_hook` which would be faster, but I'd like to add `assert(_PyType_Lookup(type, &_Py_ID(__getattr__)) == NULL)` in that case.

Comment (Fidget-Spinner) [reply Python/specialize.c]: Thanks for providing the code. This was really helpful.

File: Python/specialize.c [hunk 15]
```diff
@@ -554,21 +559,56 @@ typedef enum {
...
+            /* Normal attribute lookup; */
+            has_getattr = false;
+        }
+        else if (getattro_slot == _Py_slot_tp_getattr_hook ||
+            getattro_slot == _Py_slot_tp_getattro) {
+            /* One or both of __getattribute__ or __getattr__ may have been
+             overridden See typeobject.c for why these functions are special. */
+            PyObject *getattribute = _PyType_Lookup(type,
+                &_Py_ID(__getattribute__));
+            PyInterpreterState *interp = _PyInterpreterState_GET();
```
Comment (kumaraditya303) [inline Python/specialize.c:589]: Not for this PR: Is there a reason that the callable_cache is per interpreter rather than global? Static types are shared by all interpreters so all interpreters have the same  slot wrappers.

**Comment by Fidget-Spinner** [comment]
Ping @markshannon . Is there anything left for me to do?

File: Python/specialize.c [hunk 16]
```diff
@@ -729,24 +771,13 @@ _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name)
                 SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_EXPECTED_ERROR);
                 goto fail;
             }
-            if (Py_TYPE(fget) != &PyFunction_Type) {
-                SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_ATTR_PROPERTY);
+            if (!Py_IS_TYPE(fget, &PyFunction_Type)) {
+                SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_NOT_PY_FUNCTION);
```
Comment (markshannon) [inline Python/specialize.c]: This isn't going to make sense in the stats. What isn't a Py function?
Change to `SPEC_FAIL_GETATTRIBUTE_NOT_PY_FUNCTION`?

**Review by markshannon — CHANGES_REQUESTED** [review]

**Comment by markshannon** [comment]
One minor issue, and we should probably run the buildbots again.
Other than that, looks good.

File: Python/specialize.c (hunk 16 above)
Comment (Fidget-Spinner) [reply Python/specialize.c]: That's for `property` specialisations. Not getattribute.

Comment (markshannon) [reply Python/specialize.c]: Then name it `SPEC_FAIL_PROPERTY_NOT_PY_FUNCTION`?
Otherwise it shows up as "not py function" which is confusing.

File: Python/specialize.c [hunk 17]
```diff
@@ -594,7 +595,9 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
                 getattribute != interp->callable_cache.object__getattribute__;
             has_getattr = _PyType_Lookup(type, &_Py_ID(__getattr__)) != NULL;
             if (has_custom_getattribute) {
-                if (!has_getattr && Py_IS_TYPE(getattribute, &PyFunction_Type)) {
+                if (getattro_slot == _Py_slot_tp_getattro &&
```
Comment (markshannon) [inline Python/specialize.c:598]: Why this change?

Comment (Fidget-Spinner) [reply Python/specialize.c:598]: An assertion about which slot is present during specialisation was failing in one of the buildbots. Apparently there are cases where the other slot (which potentially calls `__getattr__`) still stuck around even with no `__getattr__`. This may mean the slot had not been called yet (seems strange considering specialised code is supposed to be hot).

Comment (markshannon) [reply Python/specialize.c:598]: Which assertion?
Does that mean the assertion is incorrect, or that this test was?

Comment (Fidget-Spinner) [reply Python/specialize.c:598]: This one https://github.com/python/cpython/pull/93988/files#diff-22e653d878778ca28317261f7826545288d02fd9980cf485d0329f28a34993f2R834.

Comment (markshannon) [reply Python/specialize.c:598]: Strictly, that assert is wrong, in the sense that it's OK if `type->tp_getattro == _Py_slot_tp_getattr_hook` provided that there it would convert itself to `_Py_slot_tp_getattro`.

Having said that, it is probably easier to reason about if we assert `type->tp_getattro == _Py_slot_tp_getattro`, and the impact on performance of the extra test will be negligible.

**Review by markshannon — APPROVED** [review]

**Comment by gvanrossum** [comment]
Hey @Fidget-Spinner and @markshannon, I just noticed that `LOAD_ATTR_GETATTRIBUTE_OVERRIDDEN`, introduced here to support `__getattribute__`, is never used in the pyperformance benchmark suite. Is that specialization at all important? I know `__getattr__` is important -- but `__getattribute__` does seem like something rare enough that we might not bother. Does either of you recall why it's needed?

**Comment by markshannon** [comment]
FTR (from a discussion on discord) `__getattribute__` was used in `_DeprecatedType` in the typing module, but not since 3.12.

I think the specialization is worth keeping, but we could move into the "long tail" discussed in https://github.com/faster-cpython/ideas/issues/447
